### ISSUE TITLE
Cannot run sh_test tests on Windows [#6622 will fix it]

### ISSUE THREAD

**Issue by vmax** [bazelbuild/bazel#7387]
### Description of the problem / feature request:

I cannot `bazel run` `sh_test` on Windows whereas `bazel test` succeeds

### Feature requests: what underlying problem are you trying to solve with this feature?

As a user of `bazel` on Mac I would expect `bazel run` on a test target to behave the same as `bazel test --test_output=streamed` which doesn't seem to be the case.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Create an `sh_test` with whatever contents

Resulting error:
```
Executing tests from //:a                                                                                                                                            
external/bazel_tools/tools/test/test-setup.sh: line 147: cd: /c/users/circleci/_bazel~1/q3reexpr/execroot/__main__/bazel-out/x64_windows-fastbuild/bin/a.exe.runfiles
/__main__: No such file or directory                                                                                                                                 
Could not chdir /c/users/circleci/_bazel~1/q3reexpr/execroot/__main__/bazel-out/x64_windows-fastbuild/bin/a.exe.runfiles/__main__  
```

### What operating system are you running Bazel on?

Windows Server 2019

### What's the output of `bazel info release`?

```
 bazel version                                                                                                                                           
WARNING: --batch mode is deprecated. Please instead explicitly shut down your Bazel server using the command "bazel shutdown".                            
INFO: Invocation ID: aa5a2fea-ff69-4f76-b963-f7fb4f427831                                                                                                 
Build label: 0.22.0                                                                                                                                       
Build target: bazel-out/x64_windows-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar                                      
Build time: Mon Jan 28 12:59:27 2019 (1548680367)                                                                                                         
Build timestamp: 1548680367                                                                                                                               
Build timestamp as int: 1548680367
```

###  Have you found anything relevant by searching the web?

—

### Any other information, logs, or outputs that you want to share?

Let me know if I'm wrong in understanding `bazel run` vs `bazel test`

**Comment by laszlocsomor** [comment]
The culprit is, `bazel run` only completes a build phase while `bazel test` seems to do a bit more, namely creating the root of the runfiles tree in `bazel-bin/foo/my_test.exe.runfiles/__main__`.

When the `test-setup.sh` executes this line [1], it fails with `run`/`build` but works with `test`.

Another way to view the bug is, test-setup.sh should not try to `cd` into the runfiles directory on Windows since it's empty anyway. (On other platforms it contains the runfiles tree.)

`run`  works when the experimental native test wrapper is enabled [2]. My goal is to eventually (~in a few months) enable this by default and disable test-setup.sh on Windows, but we aren't ready for that yet.

[1] https://github.com/bazelbuild/bazel/blob/5edfa14ea4705dd4709e62e5d02d560ad4637b02/tools/test/test-setup.sh#L147

[2] follow the journey on https://github.com/bazelbuild/bazel/issues/5508

**Comment by laszlocsomor** [comment]
Fixing test-setup.sh to not fail when `$DIR` does not exist seems like a reasonable short-term workaround to me.

**Comment by laszlocsomor** [comment]
Here's what happens.

On Linux, "--enable_runfiles" is true by default, so Bazel in SymlinkTreeHelper runs the if-branch [1], and runs "build-runfiles" which builds `foo.runfiles/__main__/...<symlink tree>...` as part of building the target.

The "test" command's code path is different: it creates the test's working directory before running the test [2]. TestStrategy.java and test-setup.sh happen to both treat `<target>.runfiles/__main__` as the test's working directory -- definitely on purpose, though this agreement isn't codified anywhere.

On Windows, "--enable_runfiles" is false by default, so the else-branch runs here [3] (symlinkTreeRoot is `<target>.runfiles`), not calling "build-runfiles.exe", so only `<target>.runfiles/` (and not `<target>.runfiles/__main__`) is created. Therefore the "build" and "run" commands do not create `__main__`. The "test" command follows the same code path as on Linux, so it does create `__main__`.

[1] https://github.com/bazelbuild/bazel/blob/9d9caddac937c923dadc54d5c8a5477aa291afff/src/main/java/com/google/devtools/build/lib/exec/SymlinkTreeHelper.java#L117

[2] https://github.com/bazelbuild/bazel/blob/9d9caddac937c923dadc54d5c8a5477aa291afff/src/main/java/com/google/devtools/build/lib/exec/TestStrategy.java#L74

[3] https://github.com/bazelbuild/bazel/blob/9d9caddac937c923dadc54d5c8a5477aa291afff/src/main/java/com/google/devtools/build/lib/exec/SymlinkTreeHelper.java#L128

**Comment by vmax** [comment]
@laszlocsomor thank you for your reply! I'll stick to using `bazel test` for now then. However, I think it'll be useful to have consistent behaviour between all platforms

**Comment by laszlocsomor** [comment]
@vmax : Sounds good. For the record, `bazel run` isn't meant to be used with tests because it does not set up the same environment as `bazel test` does -- I'm surprised it works most of the time.

**Comment by laszlocsomor** [comment]
Update: I don't plan to fix this bug. It will resolve itself with https://github.com/bazelbuild/bazel/issues/6622.

With Bazel 0.26.0 and `--incompatible_windows_native_test_wrapper`, I can `bazel run` a sh_test target.

**Comment by laszlocsomor** [comment]
Fixed since Bazel 1.0 resolved #6622.
